pkg/build/strategies/dockerfile/dockerfile.go

```diff
@@ -154,41 +153,31 @@ func (builder *Dockerfile) CreateDockerfile(config *api.Config) error {
 			buffer.WriteString(fmt.Sprintf("# Save-artifacts script sourced from builder image based on user input or image metadata.\n"))
 			artifactsScript = sanitize(filepath.ToSlash(filepath.Join(imageScriptsDir, "save-artifacts")))
 		}
+		// Switch to the image user if it is not root
+		if len(imageUser) > 0 && imageUser != "root" {
+			buffer.WriteString(fmt.Sprintf("USER %s\n", imageUser))
+		}
 		buffer.WriteString(fmt.Sprintf("RUN if [ -s %[1]s ]; then %[1]s > %[2]s; else touch %[2]s; fi\n", artifactsScript, artifactsTar))
 	}
 
 	buffer.WriteString(fmt.Sprintf("FROM %s\n", config.BuilderImage))
+	// Run as root to COPY and chown source content
+	buffer.WriteString("USER root\n")
 
-	if len(config.AssembleUser) > 0 {
-		buffer.WriteString(fmt.Sprintf("USER %s\n", imageUser))
+	imageLabels := util.GenerateOutputImageLabels(builder.sourceInfo, config)
+	for k, v := range config.Labels {
+		imageLabels[k] = v
 	}
-
-	if config.Incremental {
-		buffer.WriteString(fmt.Sprintf("COPY --from=cached %[1]s %[1]s\n", artifactsTar))
-		buffer.WriteString(fmt.Sprintf("RUN chown %s:0 %s && \\\n", sanitize(imageUser), artifactsTar))
-		buffer.WriteString(fmt.Sprintf("    if [ -s %[1]s ]; then mkdir -p %[2]s; tar -xf %[1]s -C %[2]s; fi && \\\n", artifactsTar, sanitize(filepath.ToSlash(artifactsDestDir))))
-		buffer.WriteString(fmt.Sprintf("    rm %s\n", artifactsTar))
-	}
-
-	generatedLabels := util.GenerateOutputImageLabels(builder.sourceInfo, config)
-	if len(generatedLabels) > 0 || len(config.Labels) > 0 {
+	if len(imageLabels) > 0 {
 		first := true
 		buffer.WriteString("LABEL ")
-		for k, v := range generatedLabels {
+		for k, v := range imageLabels {
 			if !first {
 				buffer.WriteString(fmt.Sprintf(" \\\n      "))
 			}
 			buffer.WriteString(fmt.Sprintf("%q=%q", k, v))
 			first = false
 		}
-		for k, v := range config.Labels {
-			if !first {
-				buffer.WriteString(fmt.Sprintf(" \\\n      "))
-			}
-			buffer.WriteString(fmt.Sprintf("%q=%q", k, v))
-			first = false
-		}
-
```

Comment:
Noted that if `config.Labels` had duplicate keys, we would have the same ENV variable set twice.

Reply:
you mean the same LABEL twice?

that's ok, the latter takes precedence i believe, so as long as we set the user's value last so they can overwrite our generated label, it's fine.

Reply:
debating whether it's better as it was so we can see the values we generated in the dockerfile, or better the way you have it..   i guess i can go either way, so let's go with this.  less layers anyway.